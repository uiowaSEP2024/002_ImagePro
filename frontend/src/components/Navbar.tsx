import { useAuthContext } from "@/hooks/useAuthContext";
import { ReactNode } from "react";
import {
  Button,
  Center,
  Container,
  Flex,
  HStack,
  Link
} from "@chakra-ui/react";
import React from "react";

import NextLink from "next/link";

const NavLink = ({ children, link }: { children: ReactNode; link: string }) => (
  <Link
    as={NextLink}
    px={4}
    py={2}
    rounded={"md"}
    _hover={{
      textDecoration: "none",
      bg: "gray.100"
    }}
    href={link}
    fontWeight={"bold"}
  >
    {children}
  </Link>
);

const Navbar = () => {
  const { currentUser, logOut } = useAuthContext();

  const linksRightUnauthenticated = [
    {
      to: "/login",
      name: "Log in",
      id: "loginButton",
      show: true
    },
    {
      to: "/signup",
      name: "Sign up",
      id: "signup",
      show: true
    }
  ];

  const linksLeftUnauthenticated = [
    {
      to: "/",
      name: "Home",
      id: "home",
      show: true
    }
  ];

  const linksLeftAuthenticated = [
    {
      to: "/",
      name: "Home",
      id: "home",
      show: !!currentUser
    },
    {
      to: "/dashboard",
      name: "Dashboard",
      id: "dashboard",
      show: !!currentUser
    },
    {
      to: "/billing",
      name: "Billing",
      id: "billing",
      show: currentUser?.role === "provider"
    },
    {
      to: "/reporting",
      name: "Reporting",
      id: "reporting",
      show: currentUser?.role === "provider"
    },
    {
      to: "/apikeys",
      name: "Generate API Keys",
      id: "apikeys",
      show: currentUser?.role === "provider"
    }
  ];

  const linksRightAuthenticated = [
    {
      to: "/profile",
      name: "My Profile",
      id: "profile",
      show: !!currentUser
    },
    {
      onClick: logOut,
      name: "Logout",
      id: "logoutButton",
      show: !!currentUser
    }
  ];

  const linksRight = currentUser
    ? linksRightAuthenticated
    : linksRightUnauthenticated;

  const linksLeft = currentUser
    ? linksLeftAuthenticated
    : linksLeftUnauthenticated;

  return (
    <Center boxShadow={"xs"}>
      <Container py={2} maxW={"container.xl"} px={4}>
        <Flex alignItems={"center"} justifyContent={"space-between"}>
          <HStack as={"nav"} spacing={4} display={{ base: "none", md: "flex" }}>
            {linksLeft.map((link) => {
              if ("show" in link && !link.show) {
                return null;
              }
              return (
                <NavLink data-testid={link.id} key={link.name} link={link.to}>
                  {link.name}
                </NavLink>
              );
            })}
          </HStack>

          <HStack as={"nav"} spacing={2} display={{ base: "none", md: "flex" }}>
            {linksRight.map((link) => {
              if ("onClick" in link) {
                return (
                  <Button
                    data-testid={link.id}
                    key={link.name}
                    onClick={link.onClick}
                  >
                    {link.name}
                  </Button>
                );
              }

              return (
                <NavLink data-testid={link.id} key={link.name} link={link.to}>
                  {link.name}
                </NavLink>
              );
            })}
          </HStack>
        </Flex>
      </Container>
    </Center>
  );
};

export default Navbar;
