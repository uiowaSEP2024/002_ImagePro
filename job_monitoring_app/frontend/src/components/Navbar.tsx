import { useAuthContext } from "@/hooks/useAuthContext";
import { ReactNode } from "react";
import {
  Button,
  Center,
  Container,
  Flex,
  HStack,
  Link
} from "@chakra-ui/react";
import React from "react";

import NextLink from "next/link";

/**
 * NavLink is a functional component that renders a navigation link.
 *
 * @param {object} props - The properties passed to the component.
 * @param {ReactNode} props.children - The child elements to be rendered within the link.
 * @param {string} props.link - The URL that the link points to.
 * @returns {JSX.Element} The NavLink component.
 */
const NavLink = ({ children, link }: { children: ReactNode; link: string }): JSX.Element => (
  <Link
    as={NextLink}
    px={4}
    py={2}
    rounded={"md"}
    _hover={{
      textDecoration: "none",
      bg: "gray.100"
    }}
    href={link}
    fontWeight={"bold"}
  >
    {children}
  </Link>
);

/**
 * Navbar is a functional component that renders a navigation bar.
 * The navigation bar includes links to different pages of the application.
 * The links displayed in the navigation bar depend on whether the user is authenticated and their role.
 *
 * @returns {JSX.Element} The Navbar component.
 */
const Navbar = (): JSX.Element => {
  const { currentUser, logOut } = useAuthContext();

  // Define the links to be displayed in the navigation bar when the user is not authenticated.
  const linksRightUnauthenticated = [
    {
      to: "/login",
      name: "Log in",
      id: "loginButton",
      show: true
    },
    {
      to: "/signup",
      name: "Sign up",
      id: "signup",
      show: true
    }
  ];

  const linksLeftUnauthenticated = [
    {
      to: "/",
      name: "Home",
      id: "home",
      show: true
    }
  ];

  // Define the links to be displayed in the navigation bar when the user is authenticated.
  const linksLeftAuthenticated = [
    {
      to: "/",
      name: "Home",
      id: "home",
      show: !!currentUser
    },
    {
      to: "/dashboard",
      name: "Dashboard",
      id: "dashboard",
      show: !!currentUser
    },
    {
      to: "/jobs",
      name: "Jobs",
      id: "jobs",
      show: !!currentUser
    },
    {
      to: "/analytics",
      name: "Analytics",
      id: "analytics",
      show: currentUser?.role === "provider"
    },
    {
      to: "/reporting",
      name: "Reporting",
      id: "reporting",
      show: currentUser?.role === "provider"
    },
    {
      to: "/apikeys",
      name: "Generate API Keys",
      id: "apikeys",
      show: currentUser?.role === "provider"
    }
  ];

  const linksRightAuthenticated = [
    {
      to: "/profile",
      name: "My Profile",
      id: "profile",
      show: !!currentUser
    },
    {
      onClick: logOut,
      name: "Logout",
      id: "logoutButton",
      show: !!currentUser
    }
  ];

  // Determine which set of links to use based on whether the user is authenticated.
  const linksRight = currentUser
    ? linksRightAuthenticated
    : linksRightUnauthenticated;

  const linksLeft = currentUser
    ? linksLeftAuthenticated
    : linksLeftUnauthenticated;

  // Determine the color of the navigation bar based on the user's role.
  const navBarColor = currentUser?.role == "provider" ? "gray.100" : undefined;

  // Render the Navbar component.
  return (
    <Center bgColor={navBarColor} boxShadow={"xs"}>
      <Container py={2} maxW={"container.xl"} px={4}>
        <Flex alignItems={"center"} justifyContent={"space-between"}>
          <HStack as={"nav"} spacing={4} display={{ base: "none", md: "flex" }}>
            {linksLeft.map((link) => {
              if ("show" in link && !link.show) {
                return null;
              }
              return (
                <NavLink data-testid={link.id} key={link.name} link={link.to}>
                  {link.name}
                </NavLink>
              );
            })}
          </HStack>

          <HStack as={"nav"} spacing={2} display={{ base: "none", md: "flex" }}>
            {linksRight.map((link) => {
              if ("onClick" in link) {
                return (
                  <Button
                    data-testid={link.id}
                    key={link.name}
                    onClick={link.onClick}
                  >
                    {link.name}
                  </Button>
                );
              }

              return (
                <NavLink data-testid={link.id} key={link.name} link={link.to}>
                  {link.name}
                </NavLink>
              );
            })}
          </HStack>
        </Flex>
      </Container>
    </Center>
  );
};

export default Navbar;
