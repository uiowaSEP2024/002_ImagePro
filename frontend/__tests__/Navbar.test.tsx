// __tests__/navbar.test.jsx

import { render, screen, RenderResult, waitFor, fireEvent } from "@testing-library/react";
import Navbar from "../src/components/Navbar";
import "@testing-library/jest-dom";
import { useRouter } from "next/router";

jest.mock('next/router', () => ({
  useRouter() {
    return ({
      route: '/',
      pathname: '',
      query: '',
      asPath: '',
      push: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn()
      },  beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null)
    });
  },
}));

jest.mock("@/utils/auth", () => ({
  checkUserLoggedIn() {
    return new Promise((resolve) => {
      resolve( {detail : "already logged in!"} )
    });
  },
  sendLogOutReq(){
    console.log("Heyyy there")
    return new Promise((resolve) => {
      const data = {
        email: "user@example.com",
        first_name: "string",
        last_name: "string",
        id: 0
      }
      resolve( data)
    });
  }
}))

const router = useRouter()

describe("NavBar", () => {

  it("renders text", async () => {
  
    render(<Navbar />);

    expect(router.push).not.toBeCalledWith("/");

    const text = await waitFor(() =>
      screen.getByRole("link", {
        name: /Home/i,
      }));

    await waitFor(()=>{
      expect(text).toBeInTheDocument();
    })
  });
  
  it("logs out", async () => {
    const button = await waitFor(() => screen.getByTestId("logoutButton"));
    fireEvent.click(button);

    expect(router.push).toBeCalledWith("/");

  });

});
