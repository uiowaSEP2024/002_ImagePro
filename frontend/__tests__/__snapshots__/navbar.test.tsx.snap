// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar renders navbar unchanged 1`] = `
<div>
  <nav
    class="nextui-c-hhqfap nextui-c-hhqfap-ehawcb-variant-sticky nextui-c-hhqfap-beFEYN-maxWidth-lg nextui-c-hhqfap-kAwfsS-disableShadow-false nextui-c-hhqfap-gqZKcP-disableBlur-false nextui-c-hhqfap-iPJLV-css nextui-navbar"
  >
    <div
      class="nextui-c-egxOLa nextui-navbar-container"
    >
      <ul
        class="nextui-c-jklhtj nextui-c-dGYPDG nextui-c-PJLV nextui-c-dGYPDG-hHixqa-enableCursorHighlight-true nextui-c-PJLV-ghzjVU-hideIn-xs nextui-c-jklhtj-ijFvnTw-css nextui-navbar-content"
        style="--nextui--transitionDuration: 0ms; --nextui--opacity: 0; --nextui--width: 0px; --nextui--left: calc(var(--nextui--navbarContentItemHorizontalPadding) * 0.5 * -1);"
      >
        <li
          class="nextui-c-kOXsPe nextui-c-dAPDu nextui-c-dAPDu-kJGdUE-color-default nextui-c-kOXsPe-iPJLV-css nextui-navbar-cursor-highlight"
        />
        <li
          class="nextui-c-kOXsPe nextui-c-yrlUe nextui-c-PJLV nextui-c-yrlUe-kJGdUE-activeColor-default nextui-c-yrlUe-cdrGrt-isActive-true nextui-c-yrlUe-hYJzFf-underlineHeight-normal nextui-c-kOXsPe-ikPzmPl-css nextui-navbar-item"
          id="react-aria4616176951-1"
        >
          <a
            class="nextui-c-kzikkm nextui-c-gulvcB nextui-c-kzikkm-hfnKRU-color-default nextui-c-kzikkm-isWIbj-animated-true nextui-c-kzikkm-iejNGLI-css nextui-link nextui-navbar-link"
            href="/"
            id="react-aria4616176951-1"
          >
            Home
          </a>
        </li>
        <li
          class="nextui-c-kOXsPe nextui-c-yrlUe nextui-c-PJLV nextui-c-yrlUe-kJGdUE-activeColor-default nextui-c-yrlUe-hYJzFf-underlineHeight-normal nextui-c-kOXsPe-ifzFbZK-css nextui-navbar-item"
          id="react-aria4616176951-2"
        >
          <a
            class="nextui-c-kzikkm nextui-c-gulvcB nextui-c-kzikkm-hfnKRU-color-default nextui-c-kzikkm-isWIbj-animated-true nextui-c-kzikkm-iejNGLI-css nextui-link nextui-navbar-link"
            href="#"
            id="react-aria4616176951-2"
          >
            Dashboard
          </a>
        </li>
        <li
          class="nextui-c-kOXsPe nextui-c-yrlUe nextui-c-PJLV nextui-c-yrlUe-kJGdUE-activeColor-default nextui-c-yrlUe-hYJzFf-underlineHeight-normal nextui-c-kOXsPe-ifzFbZK-css nextui-navbar-item"
          id="react-aria4616176951-3"
        >
          <a
            class="nextui-c-kzikkm nextui-c-gulvcB nextui-c-kzikkm-hfnKRU-color-default nextui-c-kzikkm-isWIbj-animated-true nextui-c-kzikkm-iejNGLI-css nextui-link nextui-navbar-link"
            href="#"
            id="react-aria4616176951-3"
          >
            Billing
          </a>
        </li>
      </ul>
      <ul
        class="nextui-c-jklhtj nextui-c-dGYPDG nextui-c-PJLV nextui-c-dGYPDG-hHixqa-enableCursorHighlight-true nextui-c-PJLV-ghzjVU-hideIn-xs nextui-c-jklhtj-ijFvnTw-css nextui-navbar-content"
        style="--nextui--transitionDuration: 0ms; --nextui--opacity: 0; --nextui--width: 0px; --nextui--left: calc(var(--nextui--navbarContentItemHorizontalPadding) * 0.5 * -1);"
      >
        <li
          class="nextui-c-kOXsPe nextui-c-dAPDu nextui-c-dAPDu-kJGdUE-color-default nextui-c-kOXsPe-iPJLV-css nextui-navbar-cursor-highlight"
        />
        <li
          class="nextui-c-kOXsPe nextui-c-yrlUe nextui-c-PJLV nextui-c-yrlUe-kJGdUE-activeColor-default nextui-c-yrlUe-hYJzFf-underlineHeight-normal nextui-c-kOXsPe-ifzFbZK-css nextui-navbar-item"
          id="react-aria4616176951-4"
        >
          <a
            class="nextui-c-kzikkm nextui-c-gulvcB nextui-c-kzikkm-hfnKRU-color-default nextui-c-kzikkm-isWIbj-animated-true nextui-c-kzikkm-iejNGLI-css nextui-link nextui-navbar-link"
            href="#"
            id="react-aria4616176951-4"
          >
            Login
          </a>
        </li>
        <li
          class="nextui-c-kOXsPe nextui-c-yrlUe nextui-c-PJLV nextui-c-yrlUe-kJGdUE-activeColor-default nextui-c-yrlUe-cdrGrt-isActive-true nextui-c-yrlUe-hYJzFf-underlineHeight-normal nextui-c-kOXsPe-ikPzmPl-css nextui-navbar-item"
          id="react-aria4616176951-5"
        >
          <a
            class="nextui-c-kzikkm nextui-c-gulvcB nextui-c-kzikkm-hfnKRU-color-default nextui-c-kzikkm-isWIbj-animated-true nextui-c-kzikkm-iejNGLI-css nextui-link nextui-navbar-link"
            href="/signup"
            id="react-aria4616176951-5"
          >
            Sign up
          </a>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;
